### ISSUE TITLE
Pamauthentication and normalization

### ISSUE THREAD

**Issue by chrlunden** [jupyterhub/jupyterhub#2386]
Hi,

my very first post on github ever, so be gentle :)

My problem is related to authentication against a AD-LDS system with objects that are manipulated to work with our Linux Servers. However, due to several merges, we have encountered a few problems with users that have been allowed to created uppercase usernames. 

It worked like a charm before one of those user tried to log in. 

Could anyone give me a little help in how I can allow any kind of username to log in and create a home catalog on our systems? for instance "BERNIE"

I have tried to look into auth.py and comment out:
username = username.lower() 

But I am quite sure that there are several files that needs convincing ? 

Thanks in advance for any help regarding this issue.

regards,

Christian

**Comment by Zsailer** [comment]
This is related to #1671.

**Comment by Zsailer** [comment]
I think your intuition is correct, but the problem may be a bit deeper. 

I'm not 100% sure why the default behavior is to normalize Hub usernames to lowercase (@minrk would be able to explain better). I'm guessing that this was a deliberate choice because Jupyterhub could be deployed on both case sensitive/insensitive backends. Normalizing to lowercase forces Hub admins to enforce such a rule for users. 

You can change this default behavior by overriding the `normalize_username` method (similar to what you did by commenting out). Though, the preferred approach would be to subclass the PAMAuthenticator. You can do this directly in your `jupyterhub_config.py` file.  It would look something like this:

```python
from jupyterhub.auth import PAMAuthenticator

class MyPAMAuthenticator(PAMAuthenticator):

    def normalize_username(self, username):
        username = self.username_map.get(username, username)
        return username

c.JupyterHub.authenticator_class = MyPAMAuthenticator
```

**Comment by Zsailer** [comment]
>Our ID-section is looking into ways to alter all objects containing upper case user names.

That sounds like a pain! 

In #1671, there's talk about adding a "denormalize" mechanism to the Hub that would solve this issue. Then, you won't have to mess with changing the usernames you inherited from the merge. 

I'll follow up on that and let you know!

**Comment by Zsailer** [comment]
Check out #2397. Someone beat me to the PR. 😃

**Comment by minrk** [comment]
Closed by #2397
